### DISCUSSION TITLE
How do I avoid casts when using `DefaultArgs`?

### DISCUSSION THREAD

**Discussion by divmgl** [prisma/prisma#24867] [Q&A]
### Question

I have this payload:

```tsx
export const UNIFIED_APPLICATION_PAYLOAD =
  Prisma.validator<Prisma.unified_applicationDefaultArgs>()({
    select: {
      // some fields here
    }
  })
```

I want to be able to pass it into any Prisma call, but I'm having to manually cast in order for the typing to work:

```tsx
  public forProject<T extends Prisma.unified_applicationDefaultArgs>(
    projectId: string,
    payload?: T,
    txn: Prisma.TransactionClient = this.prisma
  ) {
    return txn.unified_application.findMany({ where: { projectId, ...payload } }) as Promise<
      Array<Prisma.unified_applicationGetPayload<T>>
    >
  }
```

Is there any way I can avoid this `as Promise<Array<Prisma.unified_applicationGetPayload<T>>>` cast?

**Comment by ludralph** [comment]
Hi @divmgl 👋 

Can you specify the return type directly in the function signature, TypeScript can infer the correct type without needing an explicit cast. Here's how you can modify your forProject method:
```ts
import { Prisma } from '@prisma/client';

export const UNIFIED_APPLICATION_PAYLOAD = Prisma.validator<Prisma.unified_applicationDefaultArgs>()({
  select: {
    // some fields here
  }
});

public forProject<T extends Prisma.unified_applicationDefaultArgs>(
  projectId: string,
  payload?: T,
  txn: Prisma.TransactionClient = this.prisma
): Promise<Array<Prisma.unified_applicationGetPayload<T>>> {
  return txn.unified_application.findMany({ where: { projectId, ...payload } });
}
```
In this updated version, The return type of the forProject method is specified as Promise<Array<Prisma.unified_applicationGetPayload<T>>>. This allows TypeScript to infer the correct return type based on the generic type T without needing an explicit cast.

**Reply by divmgl** [reply]
Unfortunately this will still show a type error.

**Reply by divmgl** [reply]
```tsx
  async forApplication<T extends Prisma.application_activityDefaultArgs>(
    applicationId: string,
    payload?: T,
    txn: Prisma.TransactionClient = this.prisma
  ): Promise<Array<Prisma.application_activityGetPayload<T>>> {
    return await txn.application_activity.findMany({
      where: { applicationId },
      orderBy: { createdAt: "asc" },
      ...payload,
    })
  }
```

Cast is still necessary. The type error:

```
Type '{ id: string; type: application_activity_type; payload: JsonValue; createdAt: Date; deletedAt: Date | null; applicationId: string; }[]' is not assignable to type 'GetFindResult<$application_activityPayload<DefaultArgs>, T>[]'.
  Type '{ id: string; type: application_activity_type; payload: JsonValue; createdAt: Date; deletedAt: Date | null; applicationId: string; }' is not assignable to type 'GetFindResult<$application_activityPayload<DefaultArgs>, T>'.ts(2322)
```

**Reply by ludralph** [reply]
Hi @divmgl 

Can you try to explicitly type the payload parameter:
```js
public forProject<T extends Prisma.unified_applicationSelect>(
  projectId: string,
  payload?: Prisma.unified_applicationFindManyArgs,
  txn: Prisma.TransactionClient = this.prisma
): Promise<Array<Prisma.unified_applicationGetPayload<{ select: T }>>> {
  return txn.unified_application.findMany({
    where: { projectId },
    ...payload
  });
}
```
If you specifically need to use `select`, you could try:
```js
public forProject<T extends Prisma.unified_applicationSelect>(
  projectId: string,
  select: T,
  txn: Prisma.TransactionClient = this.prisma
): Promise<Array<Prisma.unified_applicationGetPayload<{ select: T }>>> {
  return txn.unified_application.findMany({
    where: { projectId },
    select,
  });
}
```

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!;
